docs/CInterop.md

```diff
@@ -0,0 +1,285 @@
+# C Interop
+
+This is an extension of what is covered in the [Language Guide](./LanguageGuide.md#c-interop).
+
+## Content
+
+- [Managed types](#managed-types)
+  - [String](#string)
+  - [Array](#array)
+- [Embedding C code in Carp](#embedding-c-code-in-carp)
+  - [`deftemplate`](#deftemplate)
+    - [`Basic example`](#basic-example)
+    - [`Generics`](#generics)
+
+## Managed types
+
+In Carp types like `String` and `Array` are _managed_ types in that they are
+allocated on the Heap and the compiler will automatically free the allocated
+memory when they go out of scope. We'll see how we can go from these managed
+type to C and back.
+
+### String
+
+To use a managed `String` with a C function requiring a `char*` you can use the
+`String.cstr` function that will turn your `(Ref String)` into `(Ptr CChar)`:
+
+
+```clojure
+(register puts (Fn [(Ptr CChar)] ()))
+
+(let [a-str @"A string."]
+  (puts (String.cstr &a-str)))
+(puts (String.cstr "Hello"))
+```
+
+You may want to hide the C type from the end-user:
+
+```clojure
+(defmodule MyMod
+  (hidden puts-c)
+  (private puts-c)
+  (register puts-c (Fn [(Ptr CChar)] ()) "puts")
```

Comment:
It would be nice to have some commentary on why we have to override the `puts-c` declaration manually here - I got into all kinds of trouble because I missed that out when registering a function like this inside a module (the generated C code ended up referencing non-existent functions).

Reply:
I mean in this specific case it's because I want the function to be called `puts-c` but I'll add a section about how identifiers gets generated.